coraza.go

```diff
@@ -79,48 +87,57 @@ func (m *corazaModule) Validate() error {
 
 // ServeHTTP implements caddyhttp.MiddlewareHandler.
 func (m corazaModule) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddyhttp.Handler) error {
-	var err error
 	id := randomString(16)
 	tx := m.waf.NewTransactionWithID(id)
 	defer func() {
 		tx.ProcessLogging()
 		_ = tx.Close()
 	}()
-	repl := r.Context().Value(caddy.ReplacerCtxKey).(*caddy.Replacer)
-	repl.Set("http.transaction_id", id)
 
-	it, err := processRequest(tx, r, m.logger)
-	if err != nil {
-		return err
-	}
-	if it != nil {
-		return interrupt(nil, tx, id)
+	// Early return, Coraza is not going to process any rule
+	if tx.IsRuleEngineOff() {
+		// response writer is not going to be wrapped, but used as-is
+		// to generate the response
+		return next.ServeHTTP(w, r)
 	}
 
-	rec := newStreamRecorder(w, tx)
-	err = next.ServeHTTP(rec, r)
-	if err != nil {
-		return err
-	}
-	// If the response was interrupted during phase 3 or 4 we can stop the response
-	if tx.IsInterrupted() {
-		return interrupt(nil, tx, id)
-	}
-	if !rec.Buffered() {
-		// Nothing to do, response was already sent to the client
+	repl := r.Context().Value(caddy.ReplacerCtxKey).(*caddy.Replacer)
+	repl.Set("http.transaction_id", id)
+
+	// ProcessRequest is just a wrapper around ProcessConnection, ProcessURI,
+	// ProcessRequestHeaders and ProcessRequestBody.
+	// It fails if any of these functions returns an error and it stops on interruption.
+	if it, err := processRequest(tx, r); err != nil {
+		return caddyhttp.HandlerError{
+			StatusCode: 500,
+			ID:         tx.ID(),
+			Err:        err,
+		}
+	} else if it != nil {
+		w.WriteHeader(obtainStatusCodeFromInterruptionOrDefault(it, http.StatusOK))
```

Comment:
Small doubt that can be extended to also the Coraza middleware: why are we defaulting to `http.StatusOK`? I see something like `StatusForbidden` as a better fit considering that an interruption has been raised 🤔